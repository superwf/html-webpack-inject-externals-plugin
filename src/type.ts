import type { HtmlTagObject } from 'html-webpack-plugin'

/** 每个外部模块的加载配置 */
export type PackageOption = {
  host?: string
  name: string
  path?: string
  /**
   * 使用fullPath则完全使用该项，不再从模块的package.json中自动拼接路径
   * */
  fullPath?: string
  attributes?: Record<string, string | boolean>
  /**
   * 默认按path或fullPath中的后缀名判断，js为script，css为link
   * 没有后缀默认按script
   * 也可以使用该项指定，覆盖自动判断行为
   * */
  tagName?: string

  /**
   * 每个标签前面面，可以自定义一个跟随的标签
   * 内容完全自定义
   * */
  injectBefore?: HtmlTagObject

  /**
   * 每个标签后面，可以自定义一个跟随的标签
   * 内容完全自定义
   * */
  injectAfter?: HtmlTagObject

  /**
   * 是否采用本地模式，即将node_modules中的文件复制到发布文件夹中
   * 使用了fullPath的package不会处理
   * */
  local?: boolean
}

/** 当前plugin的配置参数 */
export interface PluginOption {
  host?: string
  packages: PackageOption[]
  attributes?: Record<string, string | boolean>
  local?: boolean
}

export type PackageTagAttribute = {
  url: string
  version?: string
  attributes?: Record<string, string | boolean>
  injectBefore?: HtmlTagObject
  injectAfter?: HtmlTagObject
}
