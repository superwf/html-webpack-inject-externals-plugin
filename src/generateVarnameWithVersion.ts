/**
 * 生成变量与版本号的新合法变量名
 * */
export const generateVarnameWithVersion = (name: string, version: string) => {
  version = version.replace(/[^0-9a-z_$]/g, '')
  return `${name}${version}`
}
