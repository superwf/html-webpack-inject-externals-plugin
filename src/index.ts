/* eslint-disable import/first */
import { existsSync } from 'fs'
import { join as pathJoin } from 'path'

import chalk from 'chalk'
import cpfile from 'cp-file'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import type { HtmlTagObject } from 'html-webpack-plugin'
import urlJoin from 'url-join'
import type { Compiler, WebpackPluginInstance } from 'webpack'

import type { PackageOption, PackageTagAttribute, PluginOption, TagObject } from './type'

export class HtmlWebpackInjectExternalsPlugin implements WebpackPluginInstance {
  public name = 'HtmlWebpackInjectExternalsPlugin'

  public options: PluginOption

  constructor(options: PluginOption) {
    this.options = options
  }

  public apply(compiler: Compiler) {
    const generalHost = this.options.host
    const getBrowserFilePath = (pkg: PackageOption): PackageTagAttribute => {
      const { host, path, fullPath, name } = pkg
      if (fullPath) {
        return {
          url: fullPath,
          attributes: pkg.attributes,
          injectBefore: pkg.injectBefore,
          injectAfter: pkg.injectAfter,
        }
      }
      let pkgInfo = {} as any
      /** mono packages project need to inspact node_modules in more place */
      try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires,global-require,import/no-dynamic-require
        pkgInfo = require(`${name}/package.json`)
      } catch (e) {
        const currentPathPkgPath = pathJoin(process.cwd(), 'node_modules', `${name}/package.json`)
        if (existsSync(currentPathPkgPath)) {
          // eslint-disable-next-line @typescript-eslint/no-var-requires,global-require,import/no-dynamic-require
          pkgInfo = require(currentPathPkgPath)
        } else {
          throw new Error(`get package ${name} package.json failed`)
        }
      }
      const { version, unpkg, browser, umd } = pkgInfo
      const unpkgHost = host || generalHost
      /**
       * try to guess the browser env usage file path
       * the guess order is reference from unpkg project
       * */
      const browserFilePath = path !== undefined ? path : unpkg || browser || umd || ''
      /** 是否为本地模式 */
      const isLocal = Boolean(pkg.local !== undefined ? pkg.local : this.options.local)
      let url = ''
      if (isLocal) {
        const localPrefix = pkg.localPrefix || this.options.localPrefix || ''
        url = pathJoin('/', localPrefix, `${name}@${pkgInfo.version}/${browserFilePath}`)
        /**
         * clean-webpack-plugin will remove all copied files in done hook.
         * to work with clean-webpack-plugin, the copy task must delay to done hook
         * */
        compiler.hooks.done.tap('copy-vendor-file', () => {
          const packagePath = require.resolve(name).replace(/node_modules.+$/, `/node_modules/${name}`)
          const fromFile = pathJoin(packagePath, browserFilePath)
          const toFile = pathJoin(compiler.options.output.path || 'dist', url)
          cpfile.sync(fromFile, toFile)
          const fromFileMap = `${fromFile}.map`
          if (existsSync(fromFileMap)) {
            cpfile.sync(fromFileMap, `${toFile}.map`)
          }
        })
      } else {
        if (!unpkgHost) {
          throw new Error(`package ${name} missing host`)
        }
        url = fullPath || urlJoin(unpkgHost, `${name}@${version}`, browserFilePath)
      }
      return {
        url,
        version,
        attributes: pkg.attributes,
        injectBefore: pkg.injectBefore,
        injectAfter: pkg.injectAfter,
      }
    }
    const deps = this.options.packages.map(getBrowserFilePath)
    const tags = deps.reduce<TagObject[]>((r, d) => {
      const result: TagObject = d.url.endsWith('.css')
        ? {
            tagName: 'link',
            voidTag: true,
            attributes: {
              rel: 'stylesheet',
              href: d.url,
              ...this.options.attributes,
              ...d.attributes,
            },
          }
        : {
            tagName: 'script',
            voidTag: false,
            attributes: {
              defer: false,
              src: d.url,
              ...this.options.attributes,
              ...d.attributes,
            },
          }
      if (d.injectBefore) {
        r.push(d.injectBefore)
      }
      r.push(result)
      if (d.injectAfter) {
        r.push(d.injectAfter)
      }
      return r
    }, [])

    compiler.hooks.compilation.tap(this.name, compilation => {
      HtmlWebpackPlugin.getHooks(compilation).alterAssetTagGroups.tap(this.name, data => {
        /** remove repeat */
        const toPrependTags: any = []
        ;[...tags].reverse().forEach(tag => {
          const url = tag.tagName === 'script' ? tag.attributes.src : tag.attributes.href
          /**
           * 检查是否有重复加载的内容
           * 如果是script则对比src
           * 如果是link则对比href
           * */
          const exist = data.headTags.some(headTag => {
            if (headTag.tagName === 'script') {
              return Boolean(headTag.attributes.src) && headTag.attributes.src === url
            }
            if (headTag.tagName === 'link') {
              return headTag.attributes.href === url
            }
            return false
          })
          if (!exist) {
            data.headTags.unshift(tag as HtmlTagObject)
            toPrependTags.push(tag)
          }
        })

        console.log(chalk.rgb(73, 204, 144)(`[${this.name}] has prepend these resources in your html head:`))
        toPrependTags.forEach((tag: any) => {
          console.log(chalk.rgb(137, 191, 4)(JSON.stringify(tag)))
        })
        return data
      })
    })
  }
}
