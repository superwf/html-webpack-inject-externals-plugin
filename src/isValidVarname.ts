/** 判断字符串是否为合法变量名 */
export const isValidVarname = (name: string) => {
  return /^[a-z_$][0-9a-z_$]*?$/i.test(name)
}
