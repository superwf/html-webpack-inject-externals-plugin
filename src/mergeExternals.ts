import type { Configuration } from 'webpack'

type Externals = Configuration['externals']

/**
 * 合并webpack config externals配置
 * 返回数组化配置
 * */
export const mergeExternals = (externals1: Externals, externals2?: Externals): Externals => {
  const externals3: Externals = []
  if (externals1) {
    if (Array.isArray(externals1)) {
      externals3.push(...externals1)
    } else {
      externals3.push(externals1)
    }
  }
  if (externals2) {
    if (Array.isArray(externals2)) {
      externals3.push(...externals2)
    } else {
      externals3.push(externals2)
    }
  }
  return externals3
}
