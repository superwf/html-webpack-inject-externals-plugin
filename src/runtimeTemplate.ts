export type DEP = {
  url: string
  version?: string
  varName?: string
  attributes?: Record<string, string | boolean>
}

export const globalName = '__LOAD_EXTERNAL_DEPENDENCIES_WEBPACK_PLUGIN__'

const countTimeString = `load all dependencies by ${globalName} spend`

/** 模板头部，只在第一个入口文件中插入 */
export const templateBegin = (deps: DEP[]) => `
window.${globalName} = (function() {
  const loadedUrlSet = new Set();
  const deps = ${JSON.stringify(deps)};
  return new Promise((allResolve, allReject) => {
    const append = (index) => {
      if (index >= deps.length) {
        allResolve();
        return Promise.resolve();
      };
      const url = deps[index].url;
      if (url.endsWith('.css')) {
        const linkNode = document.createElement('link');
        linkNode.setAttribute('rel', 'stylesheet');
        linkNode.href = url;
        document.head.appendChild(linkNode);
        return append(index + 1);
      };
      if (loadedUrlSet.has(url)) {
        return append(index + 1);
      };
      const scriptNode = document.createElement('script');
      scriptNode.setAttribute('rel', 'preload');
      scriptNode.src = url;
      document.head.appendChild(scriptNode);
      const varName = deps[index].varName;
      const version = deps[index].version;
      return new Promise((resolve, reject) => {
        scriptNode.onload = () => {
          if (window[varName] && version) {
            window[varName + version.replace(/[^0-9a-z_$]/g, '')] = window[varName]
          }
          resolve();
          loadedUrlSet.add(url);
          append(index + 1);
        };
        scriptNode.onerror = () => {
          reject();
          allReject();
        };
      });
    };
    console.time('${countTimeString}');
    append(0);
  });
})().then(() => {console.timeEnd('${countTimeString}')});
window.${globalName}.documentCurrentScript = document.currentScript;
`

/** 在所有入口文件中插入 */
export const templateEnd = [
  `;${globalName}.then(() => {
`,
  `
});`,
]
