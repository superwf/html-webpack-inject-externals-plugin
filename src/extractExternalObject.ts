import { isPlainObject, reduce } from 'lodash'
// import type { ExternalsObjectElement } from 'webpack'

import { isValidVarname } from './isValidVarname'
import { Externals } from './mergeExternals'

/**
 * 提取所有对象类型的externals配置
 * 返回{[k: packageName]: globalVarname}
 * */
export const extractExternalObject = (externals: Externals): Record<string, string> => {
  return externals.reduce<Record<string, string>>((result, extern) => {
    if (isPlainObject(extern)) {
      const externObj = reduce(
        extern as Record<string, any>,
        (r, v, k) => {
          /**
           * case:
           * 'moment': 'moment'
           * */
          if (typeof v === 'string' && isValidVarname(v)) {
            r[k] = v
          }

          /**
           * case:
           * lodash : {
           *   commonjs: 'lodash',
           *   amd: 'lodash',
           *   root: '_' // indicates global variable
           * },
           * */
          if (isPlainObject(v) && v.root) {
            r[k] = v.root
          }

          /**
           * case:
           * 'antd/lib/locales/zhCN': ['antd', 'locales', 'zh_CN'],
           * */
          if (Array.isArray(v) && v.length > 0 && isValidVarname(v[0])) {
            r[k] = v[0]!
          }
          return r
        },
        {} as Record<string, string>,
      )
      return {
        ...result,
        ...externObj,
      }
    }
    return result
  }, {})
}
