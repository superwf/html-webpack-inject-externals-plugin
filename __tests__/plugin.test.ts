import * as fs from 'fs'
import * as path from 'path'

import HtmlWebpackPlugin = require('html-webpack-plugin')
import webpack from 'webpack'
import { HtmlWebpackLoadUnpkgScriptsPlugin } from '../src/index'


const resolveRoot = (str: string) => path.resolve(__dirname, '..', str)

describe('生成html注入script', () => {
  it('添加script标签到head', done => {
    webpack(
      {
        mode: 'development',
        entry: resolveRoot('__tests__/fixture/index.js'),
        output: {
          filename: '[name].js',
          // module: true,
        },
        // experiments: {
        //   outputModule: true,
        //   syncWebAssembly: true,
        //   topLevelAwait: true,
        //   asyncWebAssembly: true,
        // },
        plugins: [
          new HtmlWebpackLoadUnpkgScriptsPlugin({
            externals: {
              lodash: '_',
            },
            host: 'http://unpkg.jd.com',
            packages: [
              {
                name: 'lodash',
                path: '/lodash.js',
                attributes: {
                  type: 'module',
                },
              },
            ],
          }),
          new HtmlWebpackPlugin({
            inject: true,
            template: resolveRoot('__tests__/fixture/index.html'),
          }),
        ],
      },
      err => {
        expect(err).toBe(null)
        expect(fs.readFileSync('./dist/index.html', 'utf-8')).toMatchSnapshot()
        done()
      },
    )
  })
})
