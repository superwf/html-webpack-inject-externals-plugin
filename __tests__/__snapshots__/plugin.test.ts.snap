// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`生成html注入script 添加script标签到head 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv=\\"content-type\\" content=\\"text/html; charset=utf-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <meta name=\\"apple-mobile-web-app-capable\\" content=\\"yes\\" />
    <meta name=\\"apple-mobile-web-app-status-bar-style\\" content=\\"black\\" />
    <meta name=\\"format-detection\\" content=\\"telephone=no\\" />
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge,chrome=1\\" />
    <meta name=\\"renderer\\" content=\\"webkit\\" />
    <meta name=\\"force-rendering\\" content=\\"webkit\\" />
  <script src=\\"http://unpkg.jd.com/lodash@4.17.20/lodash.js\\" type=\\"module\\"></script></head>
  <body>
  <script src=\\"main.js\\"></script></body>
</html>
"
`;
